### PR TITLE
zig cc: Treat cu files as C++ source files

### PR DESCRIPTION
**Pull request by saierd**
First step towards #10634.

Treating stub files as C++ allows to use zig c++ as a host
compiler for nvcc.

Treating cu files as C++ allow using zig c++ as a host compiler in
CMake. CMake calls the host compiler with -E on a cu file to identify
the compiler.

Using zig c++ to directly compile CUDA code is untested.

### CODE REVIEW

**Comment by andrewrk** [comment]
Are `.cu` files and `.stub` files strictly C++ code? Or are they something different, that is also passed to a c++ compiler?

If so, we should add an additional enum tag to `FileExt` for it.

**Comment by saierd** [comment]
Thank you for merging. Will this change be included in 0.9.1 or is there already a branch for that version?

---

To answer your question:

The `.stub` files are actually C++ as far as I can tell.

`.cu` files are not really C++, they contain some [language extensions](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#c-language-extensions). As I wrote in #10634, clang can compile `.cu` files itself but this probably needs some  more work to pass the correct parameters to clang (which then requires an additional enum tag).

That is not what is usually done, though. When using Nvidia's `nvcc` for compiling a `.cu` file it splits out the C++ code into a separate file and calls a host compiler to process it, which can be `zig c++` now. See the [`nvcc` documentation](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#cuda-compilation-trajectory) for more information on how this works. In principle `zig c++` should never see the `.cu` file in this mode. When using CMake, though, it tries to identify the host compiler by executing it with `-E` on a `.cu` file directly. That is what this pull request aimed to fix as a first step.

In summary, with this pull request `zig c++` can be used as a host compiler for CUDA using CMake and `nvcc`. The direct compilation of `.cu` files using clang's CUDA mode probably does not work yet and needs some additional handling of `.cu` files.

**Comment by andrewrk** [comment]
This is eligible for being included in the bug fix release.

**Comment by gwenzek** [comment]
@saierd  If you're interested in Cuda programming with Zig you should look at #10189 and [cudaz](https://github.com/gwenzek/cudaz)

**Comment by saierd** [comment]
> @saierd If you're interested in Cuda programming with Zig you should look at #10189 and [cudaz](https://github.com/gwenzek/cudaz)

Thank you for the reference, this is interesting. I already have code in C++, though, I just want to compile it :wink:
